### PR TITLE
jazz2: switch to finalAttrs, use runCommandLocal for content, migrate to by-name

### PR DESCRIPTION
**Pull request by surfaceflinger**
## Description of changes

as in title, also 2.1.0 -> 2.2.2

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by OPNA2608 — APPROVED** [review]
Result of `nixpkgs-review pr 259774` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>jazz2</li>
    <li>jazz2-content</li>
  </ul>
</details>

Tested with shareware assets installed, still works fine (didn't really expect any effect there, but I was curious about the game :stuck_out_tongue:). Changes LGTM.

---

Some comments on `meta` things that weren't touched in this. Wouldn't say it's *necessary* to address them right here and now, but would be nice if they were fixed at some point:
- `jazz2-content` should get a `meta.platforms` attribute, `platforms.all` would work
  - Could also just do `meta = jazz2.meta // { description = "..."; }` and inherit it from `jazz2`
- `gpl3` license is deprecated, should be clarified to either `gpl3Only` or `gpl3Plus`
- `jazz2` should get an explicit `meta.mainProgram = "jazz2"`, implicit definition is deprecated

(Also IMO this package should be named `jazz2-resurrection` instead, just `jazz2` sounds too… official? iunno)

**Comment by surfaceflinger** [comment]
> (Also IMO this package should be named `jazz2-resurrection` instead, just `jazz2` sounds too… official? iunno)

I have actually talked to the upstream developer about this and he prefers all packages to be called just `jazz2`. It's already done this way in AUR, Gentoo, OpenSUSE and tracked under Repology by `jazz2`.

Regarding other suggestions - I'll handle this in a moment

File: pkgs/by-name/ja/jazz2-content/package.nix [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{ jazz2
+, lib
+, runCommandLocal
+}:
+
+runCommandLocal "jazz2-content"
+{
+  inherit (jazz2) version src;
+
+  meta = jazz2.meta // {
```
Comment (OPNA2608) [inline pkgs/by-name/ja/jazz2-content/package.nix:10]: Forgot that this now also propagates to over here, sorry.
```suggestion
  meta = (builtins.removeAttrs jazz2.meta ["mainProgram"]) // {
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Hmm… Can't say I agree with upstream, but fine.

**Review by OPNA2608 — APPROVED** [review]
LGTM, thanks!

**Review by OPNA2608 — APPROVED** [review]
Nice timing :smile:. Checked 2.2.0 on x86_64-linux, also works.

**Review by OPNA2608 — APPROVED** [review]

File: pkgs/by-name/ja/jazz2/package.nix [hunk 2]
```diff
@@ -12,36 +12,40 @@
...
-    rev = version;
-    sha256 = "nJha7+geP2Ov7ciEDzJ+XWdiF1jzv4Oeis1DwxcpJXo=";
+    rev = finalAttrs.version;
+    hash = "sha256-+N7yhp/To5Ic9Zejc/ge37kkdPdQC5h5a1FqIPRDtLQ=";
   };
 
   patches = [ ./nocontent.patch ];
 
-  buildInputs = [ libGL SDL2 zlib glew glfw openal libvorbis libopenmpt xorg.libSM xorg.libICE xorg.libXext ];
+  buildInputs = [ glew glfw libGL libopenmpt libvorbis openal SDL2 xorg.libICE xorg.libSM xorg.libXext zlib ];
```
Comment (OPNA2608) [inline pkgs/by-name/ja/jazz2/package.nix:32]: I think if the graphics backend of choice is selectable at build-time now, then only pulling in the dependencies for the selected backend makes sense?

```suggestion
  buildInputs = [
    # common stuff used by all backends
  ] ++ lib.optionals (graphicsLibrary == "GLFW") [
    # glfw-only stuff, i.e. glfw
  ] ++ lib.optionals (graphicsLibrary == "SDL2") [
    # SDL-only stuff, i.e. SDL2
  ];
```

**Review by OPNA2608 — COMMENTED** [review]
Checked SDL2 backend on Wayland, works fine.

File: pkgs/by-name/ja/jazz2/package.nix (hunk 2 above)
Comment (surfaceflinger) [reply pkgs/by-name/ja/jazz2/package.nix:32]: sure 1 sec

**Comment by surfaceflinger** [comment]
please check now, I dropped `xorg.libICE xorg.libSM xorg.libXext` from buildInputs which were required few months ago but it looks like it works now. GLEW dropped too, checked source and it's most probably used only for QT5 backend which is disabled in jazz2, just a leftover from up-upstream nCine.

**Review by OPNA2608 — APPROVED** [review]
Tested both backends on aarch64-linux w/ Wayland, LGTM!

**Comment by surfaceflinger** [comment]
> <death.killer> also I just found critical bug and I have to release version 2.2.2 tomorrow unfortunately, so if you want, you can wait with this

**Comment by surfaceflinger** [comment]
upstream added --version so added a test for it

File: pkgs/by-name/ja/jazz2/package.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    "-DLIBOPENMPT_INCLUDE_DIR=${lib.getDev libopenmpt}/include/libopenmpt"
+    "-DNCINE_DOWNLOAD_DEPENDENCIES=OFF"
+    "-DNCINE_OVERRIDE_CONTENT_PATH=${jazz2-content}"
+  ] ++ lib.optionals (graphicsLibrary == "GLFW") [
+    "-DGLFW_INCLUDE_DIR=${glfw}/include/GLFW"
+  ];
+
+  passthru.tests.version = testers.testVersion {
+    package = finalAttrs.finalPackage;
+    command = "${finalAttrs.finalPackage.meta.mainProgram} --version";
```
Comment (OPNA2608) [inline pkgs/by-name/ja/jazz2/package.nix:45]: Not necessary as that's the default. Only need to specify `command` if the program has a version flag other than `--version`, needs more setup to work in the test environment, or normally returns a non-zero status code when printing the version. Checked locally, and default works fine.

```suggestion
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 259774` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>jazz2</li>
    <li>jazz2-content</li>
  </ul>
</details>

Still works & looks good, :+1: sans version test comment.

File: pkgs/by-name/ja/jazz2/package.nix (hunk 3 above)
Comment (surfaceflinger) [reply pkgs/by-name/ja/jazz2/package.nix]: done, thanks

**Review by OPNA2608 — APPROVED** [review]

File: pkgs/by-name/ja/jazz2/package.nix [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+assert lib.assertOneOf "graphicsLibrary" graphicsLibrary [ "SDL2" "GLFW" ];
+stdenv.mkDerivation (finalAttrs: {
+  pname = "jazz2";
+  version = "2.2.2";
+
+  src = fetchFromGitHub {
+    owner = "deathkiller";
+    repo = "jazz2-native";
+    rev = finalAttrs.version;
```
Comment (SuperSandro2000) [inline pkgs/by-name/ja/jazz2/package.nix:24]: FYI this is ineffective until the hash is overwritten
